Integration of openidconnect

Fomys opened the pull request:
## Proposed change

This pull request add the support for openidconnect SSO protocol. I know it's a duplicate from #1746, but I want to update to support new version of paperless-ngx and add group support.

![image](https://github.com/paperless-ngx/paperless-ngx/assets/29522056/07f9e86a-b406-478e-95e6-2740a07b76b6)

What I plan to add with this PR:
- [x] Allow connection with openidconnect
- [x] Support groups from SSO
- [x] Allow link between SSO groups and django Groups
  - [x] API
  - [x] Frontend
- [x] Documentation
- [x] Testing

#### Current issues

- The CI fail because I create the pipfile with the wrong python version (3.11 instead of 3.8/3.9). I think this PR can wait #3761 / #3544 before merging. If it's really needed I can try to lock with the correct python version.

As it's my first contribution here and I never play with AngularJS before, I may need a careful review on some points:
- The way I create SSO groups
- API endpoint
- Frontend modifications
- Testing

### OpenIDConnect

For now, this is the only thing this PR will support (my use case), but as I use social-auth, it should be easy to integrate other authentication systems. If you want to do this, I think it's important to discuss how you can configure them (only using environment variable seems hard to maintain for more than one or two SSO systems).

### Group synchronisation

I create this PR mainly for this point: today Paperless-ngx support multi users, groups and permissions, and as most of the common SSO systems allow sharing groups, I think it can be a nice feature.

To do this, as I don't want to mess everything with existing group model, I create a new SSOGroup model, which is only a map between a Group and the name from SSO authentication.
I force synchronization of such group on each connection:
- Add user to group when needed (SSOGroup exists, user have this SSOGroup)
- Remove user from group when user loose a group (SSOGroup exists, but user don't have this SSOGroup anymore)

### Make existing application works

I want to also keep the support for existing application. I think about a system of token (already existing, but no frontend to create them, and applications listed on the wiki only support user/password) or "application password" (allow using applications, but need a lot of work to integrate in paperless-ngx).
I will do this in a second PR, because it's not related to SSO login, need a lot of work and it's a completely different feature. Please tell me if there is already a plan to extend tokens for this usage, I don't find any information on issues/PR/discussions.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (please explain):

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

PS: As this PR is at the beginning of the creation process, I won't push my change until almost everything is ok to avoid pinging everyone for every commit. Sorry for the first ping, I misunderstood the use of "draft PR".
PS2: I will do an other pull request, but I don't want to forget: settings.component.spec.ts -> test "should support delete group" use wrong variable (attempt to delete a group with the "users" variable)

Fomys committed:
OpenIDConnect backend
Use social-auth to authenticate with OpenIDConnect.
Add API to manage sso groups.

shamoon commented:
So this seems like a duplicate but not really any further than https://github.com/paperless-ngx/paperless-ngx/issues/1746 as most of the same fundamental issues need to be addressed?

Fomys commented:
It's a duplicate yes, but #1746 is for version 1.9, which lacks a lot of new features and the original author seems to have been inactive for 5 months.

I first tried copying his work and updating it to the latest version of paperless, which works, but I want to add group support. Allauth doesn't seem to support additional authentication information (such as groups, at least for openID, allauth filter only "common informations", as mail, username, picture), so I've opted for social-auth, where you can add a step to extract more information and use them.

I created the checklist from my future usages, but if you think other things are important, I can also work on it.

Fomys commented:
I just added support for groups, but without any frontend, just the api.
Right now you must create sso groups manually with the api.
I don't know if you like my solution, but as I don't want to add mess with group management, I create a new SSOGroup model to link sso group to actual django group.

shamoon commented:
Respectfully, I just think this PR is premature as it doesnt feel like this is anywhere close to ready to merge. Dont get me wrong, we appreciate the work but it's going to be pinging the group every time you make a commit, etc. Perhaps if you want to solicit help with building this you can start a discussion instead.

Your checklist seems fine but Im not sure any of us are going to be able to hop in and complete all of this work (as you can guess everyone has their own time constraints / things they're working on for the project already, and there is quite a bit here to be done).

To answer your q above: there are a few tests that might fail locally but they should actually just be skipped. I'll run ci so you can see, but again this will need to be sorted. Part of your question about how to avoid breaking changes is ensuring the tests still work. Sometimes tests will need adjustment of course but you need to make sure you can tell why. 76 sounds like too many.

Small things like adding space between buttons are trivial, so yea dont worry about those for now, that sounds like something appropriate for a draft PR, eg "this PR is ready except the following small items..."

Fomys commented:
Hi,
I know it's premature, that's why it's a draft and not a PR :sweat_smile:
I didn't know that draft PR ping everyone at each commit and is used for "almost ready" things, sorry for that. I will continue to work on it and only push to this branch when it will be in a better state. If I edit my first comment to add information about my progress, is it ok ?

My "need for help" checklist is more a "I can try something, but someone should check that I done things right", I never work with AngularJS before, and it's my first contribution here.

Thanks for the tests, I will try to solve them !

Fomys committed:
Setting page to manage SSO Groups.
Add section in "User&Group" tab to manage SSO groups.
Only display this section if user have the rights to edit it.

Fomys committed:
Update translations

Fomys committed:
Add documentation for sso configuration

Fomys commented:
Hello !
This time it's almost done, I've added the configuration page to change the sso groups, the documentation and check the tests.

I have two problems to solve before merging:
- I can't get the pipfile to lock: `Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.`, so the github action fails (like the previous ones)
- I've got an error in the "user&groups" configuration page: `Cannot find control with path: 'usersGroup -> 10'`. If I understand AngularJS correctly, this is a race condition between "add users to usersGroup" and "rendering template and list users". This error didn't exist before my modifications. If you have any tips or ideas for solving this problem, please don't hesitate to share them.

The only thing I haven't done is create tests, I'll take the time to do that later this week.

juantxorena commented:
I'm very interested in this capability, and I would like to help if needed, but I don't really know python. I just want to comment on this:

>     * I can't get the pipfile to lock: `Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.`, so the github action fails (like the previous ones) ;

I had the same problem, I tried to set up a dev environment following the [docs](https://docs.paperless-ngx.com/development/) to test the other pull request and make it work, but it doesn't work, with the same pipfile problems (something about guids).

About the other problem with angular, I'm not a frontend person neither, but I will try to look at it. But I don't have a lot of time this week or the next, so please don't count on me for fixing it.

Fomys committed:
Fix pipenv - Bump scipy version

Fomys commented:
I solve the pipenv problem by:
- Installing mysqlclient bindings
- Bumping scipy version to use python 3.11 (I see related discussion around #3761 and #3544, if I understand correctly, you plan to update scipy. I will rebase my work on this new release to allow merging)

For the front issue, I think I solve it. Can someone can look at 1aa6fcf77e5e3a4a0f34b4fbd2ffc92565cda1cb to check if my solution is correct?

shamoon commented:
Glad to see you’re making progress. I haven’t looked at this much yet. Can you explain the purpose of SSOGroup?

For the translation files you edit the source and then regenerate the translation files. Only the English file should need updating so the other changes will need to be reverted

Fomys commented:
SSOgroup is a template that links a "django" group to a group name provided by the SSO system.

I think it's better to have a separate template than to use only the name of the django group to allow finer management of groups without losing clarity.

For example, SSO provides the groups "admin", "user", "guest", Paperless contains the groups "admin", "user", "viewer", "advanced". The advantages that I see in this separate model:
- You don't have to synchronise the admins, so you can manually assign the "admin" group in paperless, without having to rename it;
- The "user" group can be synchronised automatically
- The "guest" group can be linked to "viewer".
- If the SSO also creates the "advanced" role, it will not be assigned automatically without action by the paperless administrator.

I'm forcing-pushing a clean git history, and I'm currently working on testing the frontend. Do we agree that I need to write two different series of tests: one for the "display" and one for the "service"?

Fomys committed:
Add playwright tests

sahara101 commented:
Nice, hope this feature will get released :)

Fomys committed:
Add component tests

Fomys commented:
I can officially mark this PR finished and ready !

The last commits add tests on front-end. I tried a lot, but I really don't know how to test SSO in the backend.

I force-push again to allow easy merge, I can do it again later if needed

alexgrahamuk commented:
Sometimes makeing a host file entry can help. Was recently testing a new
shib handlers with this method

A

On Wed, 2 Aug 2023, 00:04 Louis Chauvet, ***@***.***> wrote:

> I can officially mark this PR finished and ready !
>
> The last commits add tests on front-end. I tried a lot, but I really don't
> know how to test SSO in the backend.
>
> I force-push again to allow easy merge, I can do it again later if needed
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/paperless-ngx/paperless-ngx/pull/3773#issuecomment-1661217020>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOCYFQFSKW3U3ZRORD57ALXTGDOTANCNFSM6AAAAAA2DD6I2A>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

stumpylog commented on Pipfile:
```diff
@@ -59,13 +59,16 @@ zxing-cpp = {version = "*", platform_machine = "== 'x86_64'"}
 #
 # Pin this until piwheels is building 1.9 (see https://www.piwheels.org/project/scipy/)
-scipy = "==1.8.1"
+scipy = "~=1.10.0"
```
This change breaks armv7

Fomys replied:
Yes, I know, that what I explained here: https://github.com/paperless-ngx/paperless-ngx/pull/3773#issuecomment-1637147614

As you (="paperless people") already discuss about dropping support of armv7, I think it should not be a problem to wait a little bit before merging this PR (I will continue to keep it synced with dev, I use it at home)

stumpylog commented on Pipfile.lock:
```diff
@@ -51,28 +51,6 @@
             "markers": "python_full_version <= '3.11.2'",
             "version": "==4.0.2"
         },
-        "backports.zoneinfo": {
```
This change breaks Python 3.8

stumpylog commented on Pipfile:
```diff
@@ -59,13 +59,16 @@ zxing-cpp = {version = "*", platform_machine = "== 'x86_64'"}
 cryptography = "==40.0.1"
 pikepdf = "==7.2.0"
 pillow = "==9.5.0"
+# Social authentification
```
Confusingly, the correct way to add a single package, without updating many others, is `pipenv upgrade social-auth-core`.  This will not remove the libraries Python 3.8 requires.

Fomys replied:
I will try this tomorrow, with a python 3.8

Fomys replied:
After searching for a distribution where I could install 3.8, I discovered that all the major distributions already use 3.10 or 3.11 (alpine, ubuntu, debian, fedora, centos8stream, ...). And for older distribution release, most of them uses python 3.8 versions subject to CVE / already deprecated (If you want to have secure version, you should use 3.8.17, but only fedora, freebsd, gentoo and nix have this version)...
For this kind of software, where sensitive information is stored, I think it's important to keep everything up to date (especially the system itself), so you should really think about dropping armv7 support, or perhaps disabling scipy-related features (I guess "auto matching" feature?).

stumpylog replied:
These files are properly locked without breaking changes: https://gist.github.com/stumpylog/e5b7cb51ee9ae83b27e0e70e3b8362ff

shamoon commented on src-ui/e2e/settings/settings.spec.ts:
I appreciate the work that went into sorting these out, and I realize it's not totally clear at the moment because we have both e2e and unit tests but these look like they should be in unit tests. Most of the e2e tests were supplanted by the jest tests and I plan to whittle them down to only things that make sense from e2e, which would be things like using a component to change something, navigating elsewhere and verifying this.

So yea, tldr, these should be moved to jest, in which case you won't need HAR you can do things like mock permissions etc. Another minor thing is making the names even more explicit, like "should not be able to do anything" --> "should not display sso groups if insufficient permissions" or whatever.

Fomys replied:
Ho noooo, it take me whole day to figure out how to write HAR file :smile: 
I will write them in jest today. Just to be sure, it means that I have to add similar tests to src-ui/src/app/components/manage/settings/settings.component.spec.ts ?

Fomys commented:
@alexgrahamuk I don't understand what you mean by using an host file entry, can you elaborate ? And how can I integrate it with the test framework, I don't think it's a good solution to force all paperless developers to add an host entry to run tests.

shamoon commented on src-ui/src/app/components/manage/settings/settings.component.ts:
```diff
@@ -733,6 +775,68 @@ export class SettingsComponent
+  editSSOGroup(ssoGroup: PaperlessSSOGroup) {
+    var modal = this.modalService.open(SsoGroupEditDialogComponent, {
+      backdrop: 'static',
+      size: 'lg',
```
```suggestion
```

Doesnt need to be large =)

Fomys replied:
In the settings, all the modals are lg or xl, I just wanted to keep it consistent, I'll change that for my next push

shamoon replied:
Default is 'md', so all of the calls that don't specify size are smaller. Other settings have many fields as opposed to just the 2 here. Thanks

shamoon commented on src-ui/src/app/data/paperless-sso-group.ts:
```diff
@@ -0,0 +1,6 @@
+import { ObjectWithId } from './object-with-id'
+
+export interface PaperlessSSOGroup extends ObjectWithId {
+  name?: string
+  group?: number
```
```suggestion
  name: string
  group: number
```

I assume these aren't really optional

Fomys replied:
Yes, they are compulsory, but I follow what the other files do, if you look in the data folder, almost everything is optional, I thought it was normal to mark them as optional
I will fix it for my next push

shamoon commented:
Also, a general q (that I bet I could answer by reading about this more...), how does social_auth handle when a user logs in via sso what happens if the user doesnt exist in paperless?

Fomys commented:
The user is simply created, with an autogenerated username (something like `{username}{uuid}`). But the `name` field is correct, it takes the `name` field of OpenIDConnect informations.
I assume I can change this behavior (like I add group support), but I think it's a good idea to avoid conflicts with existing users.

samip5 commented:
What's the current state of this PR? It seems it's very close to being able to be merged?

Fomys commented:
For me the pr is pretty close to a finished state.

I haven't had time to finalise the requested changes yet (I'm in the middle of moving house).

I'll take the time to rebase my branch on the main repo.

I just need to understand how I can write my tests as unit tests as @shamoon asked, the things I've written are only "check that this button is displayed if permission", but for me that's only possible with playwright. If you have an example of how to do this, I'd love to see it.
It also miss some tests on "server side", but for this part it looks very complex to mock SSO, and as I use an existing SSO library, I don't think there's any point in adding a lot of complex tests.

shamoon commented:
For my part (mostly frontend):

- The e2e tests can just be removed, they are the wrong kind of test for this, they should only be jest (unit) tests
- You dont need to add testing for things like "if user doesnt have permissions, hide this component", that would be testing e.g. the *IfPermissions directive
- I think the codecov note above is inaccurate, I think your frontend coverage is better than that. Once you re-base this we'll see, frontend coverage should absolutely be 100% for this PR (its obviously largely backend), I can help if needed. There are also many tools to see this yourself e.g. running jest locally (or VSCode etc). When I look locally it looks like you do have the new (frontend) lines covered already
- There are 585 'example' jest tests in this repo =) to look at if you'd like examples (but again, I think youre ok already)

Im not sure about the best course of action regarding backend tests. I think they should be there but I do not know how they would be written.

andreheuer commented:
Is there any update on this PR? This would be a great feature and really improves the usability in small teams. Would be great to have it also in 2.0.0 :-)

shamoon committed:
Revert "Add playwright tests"

This reverts commit cfaea8ce76944bac3b654c55591a740c5053dcb1.

shamoon committed:
Re-implement frontend sso groups

shamoon committed:
Update configuration.md

shamoon committed:
Update login.html

shamoon committed:
Delete settings.spec.ts

shamoon committed:
Update translations

shamoon committed:
Update sso-group-edit-dialog.component.html

shamoon committed:
lint, fix users groups load fail

shamoon commented:
@Fomys there have been _a lot_ of changes since this PR was last worked on so I took the liberty of merging `dev` into here. The frontend should basically be OK, I may have broken the backend (Pipfile.lock merge perhaps) but I'll let you sort that out.

As for the larger questions re testing etc, Im not sure where you're at on that. But either way, I think its time to circle back to this and decide where to go from here.

shamoon committed:
Update test_management_exporter.py

shamoon committed:
Merge branch 'dev' into pr/3773

shamoon committed:
Merge context processors

shamoon committed:
Update translation strings

shamoon committed:
Basic tests for login view

shamoon committed:
Only show sso groups on frontend if enabled

shamoon committed:
Merge branch 'dev' into pr/3773

mpldr commented:
Oh god, please not stale again D:

stumpylog commented:
Unfortunately, this PR has been open for quite some time without any response from the original author.  A new attempt can always be made, but I'm going to close this zombie PR.  I do think new attempts should look at using `django-allauth` for adding OIDC support.

juantxorena commented:
You can leave it open anyway.
